psbt-xpub/src/answer.rs

```diff
@@ -20,17 +21,25 @@ pub fn add_xpub_to_psbt(
     descriptor: &Descriptor<DescriptorPublicKey>,
     psbt: &mut psbt::PartiallySignedTransaction,
 ) -> Result<(), &'static str> {
+    let secp = Secp256k1::new();
     descriptor.for_each_key(|pk| {
         if let DescriptorPublicKey::XPub(xpub) = pk.as_key() {
             let origin = match &xpub.origin {
                 Some(origin) => origin.clone(),
                 None if xpub.xkey.depth == 0 => (root_fingerprint(&xpub), vec![].into()),
                 _ => panic!("Missing key origin"),
             };
-
-            psbt.xpub.insert(xpub.xkey, origin);
+            // Check if this key matches with the PBBT-input's bip32 keymap.
+            // We need this to ensure we are not adding unrelated keys that might
+            // be present in the descriptor, into the PSBT global keymap.
+            psbt.inputs.iter().for_each(|input| {
+                input.bip32_derivation.iter().for_each(|(_, keysource)| {
```

Comment:
To support taproot descriptors you should also look at `tap_key_origins`.

Reply:
Nice idea.. will add..